### PR TITLE
Add documentation on configuration options

### PR DESCRIPTION
**Pull request by PeterKneale**
WIP - Preview of documentation
- **[more simply reviewed here than via the PR](https://github.com/PeterKneale/JustSaying/blob/feature/472-Add_documentation_relating_to_configuration/docs/Configuration.md)**
- intend on describing the various configurations
   - some I don't understand myself ie: `PointToPoint`
- diagram the resulting AWS infrastructure created
- generated using plantuml which involves 
    - running a plantuml docker container
    - installing a vscode plugin
    - you get live image rendering as you edit the puml files
    - outputs as png 

Relates to issue #472

### CODE REVIEW

File: docs/Configuration.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+# Configuring Just Saying
+
+## The publish subscribe pattern
```
Comment (martincostello) [inline docs/Configuration.md]: _"The publish-subscribe pattern"_? (dash added)

File: docs/Configuration.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+> In software architecture, publish–subscribe is a messaging pattern where senders of messages, called publishers, do not program the messages to be sent directly to specific receivers, called subscribers, but instead categorize published messages into classes without knowledge of which subscribers, if any, there may be. Similarly, subscribers express interest in one or more classes and only receive messages that are of interest, without knowledge of which publishers, if any, there are. [(wikipedia)](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern)
```
Comment (martincostello) [inline docs/Configuration.md]: Nit: _**W**ikipedia_

File: docs/Configuration.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+# Configuring a single publisher and a single subscriber
+
+## Configuring the publisher
+
+- In this configuration the following AWS resources will be created
+    - a SNS topic of name `orderplacedevent`
+    - a SQS queue of name `orderplacedevent`
+    - a SNS topic subscription for queue `orderplacedevent` to topic `orderplacedevent`
+
+- In `startup.cs` add the following
```
Comment (martincostello) [inline docs/Configuration.md]: Nit: `Startup.cs`

File: docs/Configuration.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+```cs
+    services.AddJustSaying(config =>
+    {
+        config.Publications(x =>
```
Comment (martincostello) [inline docs/Configuration.md]: In these examples I'd be tempted to use more verbose naming than `x`.

File: docs/Configuration.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        {
+            // Creates the SNS Topic
+            x.WithTopic<OrderPlacedEvent>();
+        });
+    }
+```
+
+## Configuring the subscriber
+
+- In `startup.cs` add the following
```
Comment (martincostello) [inline docs/Configuration.md]: Nit: `Startup.cs`

File: docs/Configuration.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+```cs
+    services.AddJustSaying(config =>
+    {
+        config.Subscriptions(x =>
+        {
+            // creates the SQS queue
```
Comment (martincostello) [inline docs/Configuration.md]: Nit: Sentence case

File: docs/Configuration.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            // creates the SQS to SNS subscription
+            x.ForTopic<OrderPlacedEvent>();
+        });
+    }
+    // Add a handler responsibile for handling the event
```
Comment (martincostello) [inline docs/Configuration.md]: Typo: _responsible_

File: docs/Configuration.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    services
+        .AddJustSayingHandler<OrderPlacedEvent, OrderPlacedEventHandler>();
+
+    // Add a background service that is listening for messages related to the above subscriptions
```
Comment (martincostello) [inline docs/Configuration.md]: Only one subscription here?

File: docs/Configuration.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    services.AddHostedService<Subscriber>();
+```
+
+*Resulting infrastructure*
+
+![Single Publisher - Single Subscriber](SinglePublisher_SingleSubscriber.png)
```
Comment (martincostello) [inline docs/Configuration.md]: At a later point, we'd need the diagram generation documented so we know how to regenerate it easily.

File: docs/Configuration.md (hunk 1 above)
Comment (PeterKneale) [reply docs/Configuration.md]: done

File: docs/Configuration.md (hunk 4 above)
Comment (PeterKneale) [reply docs/Configuration.md]: Not sure what to call the builders - what do you think?
```cs
    services.AddJustSaying(configureBus =>
    {
        configureBus.Subscriptions(configureSubscription =>
        {
            // Creates the SQS queue
            // Creates the SQS to SNS subscription
            configureSubscription.ForTopic<OrderPlacedEvent>();
        });
    }
```

File: docs/Configuration.md (hunk 9 above)
Comment (PeterKneale) [reply docs/Configuration.md]: See [docs/Documentation_generation.md](docs/Documentation_generation.md)

File: docs/Configuration.md (hunk 4 above)
Comment (martincostello) [reply docs/Configuration.md]: Maybe `bus` and then `subscription`, so it's like the below so the code lines read a bit like written English?

```csharp
    services.AddJustSaying(bus =>
    {
        bus.Subscriptions(subscription =>
        {
            // Creates the SQS queue
            // Creates the SQS to SNS subscription
            subscription.ForTopic<OrderPlacedEvent>();
        });
    }
```

File: docs/Configuration.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
+# Configuring Just Saying
```
Comment (slang25) [inline docs/Configuration.md]: JustSaying

File: docs/Configuration.md (hunk 4 above)
Comment (PeterKneale) [reply docs/Configuration.md]: will do - thats better - mine was too long

Comment (PeterKneale) [reply docs/Configuration.md]: done

File: docs/Configuration.md (hunk 10 above)
Comment (PeterKneale) [reply docs/Configuration.md]: done

**Comment by martincostello** [comment]
Hi @PeterKneale - what do you think there is left to do on this before it's no longer _WIP_ and we can merge it?

**Comment by martincostello** [comment]
@PeterKneale Hi Peter - are you planning to do any more on this? If not, we'll take away the WIP from the title and merge this in anyway 😃

**Comment by gkinsman** [comment]
@martincostello shall we just merge this one? 😁

**Pull request merged**